flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/security/token/HiveServer2DelegationTokenProvider.java

```diff
@@ -0,0 +1,235 @@
+         * dependencies.
+         *
+         * <p>This null check is required because the HiveServer2 provider is always on classpath
+         * but Hive jars are optional. Such case configuration is not able to be loaded. This
+         * construct is intended to be removed when HiveServer2 provider/receiver pair can be
+         * externalized (namely if a provider/receiver throws an exception then workload must be
+         * stopped).
+         */
+        if (hiveConf == null) {
+            LOG.debug(
+                    "HiveServer2 is not available (not packaged with this application), hence no "
+                            + "hiveServer2 tokens will be acquired.");
+            return false;
+        }
+        try {
+            if (!HadoopUtils.isKerberosSecurityEnabled(UserGroupInformation.getCurrentUser())) {
+                LOG.debug(
+                        "Hadoop Kerberos is not enabled,hence no hiveServer2 tokens will be acquired.");
+                return false;
+            }
+        } catch (IOException e) {
+            LOG.debug(
+                    "Hadoop Kerberos is not enabled,hence no hiveServer2 tokens will be acquired.",
+                    e);
+            return false;
+        }
+
+        if (hiveConf.getTrimmed("hive.metastore.uris", "").isEmpty()) {
+            LOG.debug(
+                    "The hive.metastore.uris item is empty,hence no hiveServer2 tokens will be acquired.");
+            return false;
+        }
+
+        if (!kerberosLoginProvider.isLoginPossible(false)) {
+            LOG.debug("Login is NOT possible,hence no hiveServer2 tokens will be acquired.");
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public ObtainedDelegationTokens obtainDelegationTokens() throws Exception {
+        UserGroupInformation freshUGI = kerberosLoginProvider.doLoginAndReturnUGI();
+        return freshUGI.doAs(
+                (PrivilegedExceptionAction<ObtainedDelegationTokens>)
+                        () -> {
+                            Preconditions.checkNotNull(hiveConf);
+                            Hive hive = Hive.get((HiveConf) hiveConf);
+                            Clock clock = Clock.systemDefaultZone();
+                            try {
+                                LOG.info("Obtaining Kerberos security token for HiveServer2");
+
+                                String principal =
+                                        hiveConf.getTrimmed(
+                                                "hive.metastore.kerberos.principal", "");
+
+                                String tokenStr =
+                                        hive.getDelegationToken(
+                                                UserGroupInformation.getCurrentUser().getUserName(),
+                                                principal);
+                                Token<HiveServer2DelegationTokenIdentifier> hive2Token =
```

Comment:
I do not get this part.
Why do we create a new Token, and why do not we use the one received from Hive?

Like:
```
                                String tokenString =
                                        client.getDelegationToken(freshUGI.getUserName(), renewer);
                                token.decodeFromUrlString(tokenString);
                                credentials.addToken(token.getKind(), token);
```

Reply:
token.decodeFromUrlString(tokenString); The token here needs to be created, otherwise the decodeFromUrlString method cannot be called